taichi/ir/type_factory.h

```diff
@@ -57,6 +60,10 @@ class TypeFactory {
   // TODO: use unordered map
   std::map<std::pair<std::string, Type *>, std::unique_ptr<Type>> tensor_types_;
 
+  std::unordered_map<std::vector<const Type *>, std::unique_ptr<Type>>
```

Comment:
Are we sure that `Type *` comparison is what we expect it to be? If not, I would be more comfortable in this case to have a custom hasher and comparison function

Reply:
Yes, `Type *` identifies a type, and two types are equal iff the `Type *`s are equal.

Reply:
If we need to add debug info (name of the elements) to it, maybe we can change to `std::vector<std::pair<const Type *, std::string>>` later.